lib/nrf_modem_lib/nrf_modem_os.c

```diff
@@ -426,7 +443,108 @@ void trace_uart_init(void)
 #endif
 }
 
-/* This function is called by nrf_modem_init and must not be called explicitly. */
+void *nrf_modem_os_alloc(size_t bytes)
+{
+	void *addr = k_heap_alloc(&library_heap, bytes, K_NO_WAIT);
+#ifdef CONFIG_NRF_MODEM_LIB_DEBUG_ALLOC
+	if (addr) {
+		LOG_INF("alloc(%d) -> %p", bytes, addr);
+	} else {
+		heap_diag.failed_allocs++;
+	}
+#endif
+	return addr;
+}
+
+void nrf_modem_os_free(void *mem)
+{
+	k_heap_free(&library_heap, mem);
+#ifdef CONFIG_NRF_MODEM_LIB_DEBUG_ALLOC
+	LOG_INF("free(%p)", mem);
+#endif
+}
+
+void *nrf_modem_os_shm_tx_alloc(size_t bytes)
+{
+	void *addr = k_heap_alloc(&shmem_heap, bytes, K_NO_WAIT);
+#ifdef CONFIG_NRF_MODEM_LIB_DEBUG_SHM_TX_ALLOC
+	if (addr) {
+		LOG_INF("shm_tx_alloc(%d) -> %p", bytes, addr);
+	} else {
+		shmem_diag.failed_allocs++;
+	}
+#endif
+	return addr;
+}
+
+void nrf_modem_os_shm_tx_free(void *mem)
+{
+	k_heap_free(&shmem_heap, mem);
+#ifdef CONFIG_NRF_MODEM_LIB_DEBUG_SHM_TX_ALLOC
+	LOG_INF("shm_tx_free(%p)", mem);
+#endif
+}
+
+void nrf_modem_lib_heap_diagnose(void)
+{
+	printk("nrf_modem heap dump:\n");
```

Comment:
Why aren't you using the logger?

Reply:
`sys_heap_dump` uses `printk` internally, so if this line used the logger the two outputs would not appear one above the other. The alternative was to enable (`select`/`imply`) `LOG_PRINTK` but I do not think this feature should require changes to the logging configuration of the application. So I sticked with `printk` here, since it's also printing in its own workqueue, it should be fine.